### PR TITLE
Added Patch to handle duplicate workflow events

### LINKED ISSUE #2310: Scheduler gets wrong update value.
STR

Create multiple schedules internal cluster and external cluster. (In my case) 
Sometimes Internal cluster workflow run gets updates it ran recently even though it was run only once. It's random behavior.

From Portal table
```
Status	  WorkflowName                             TargetCluster   ReliabilityDetails   #ofSteps    LastRun 
Succeeded argowf-chaos-kube-proxy-chaos-1604378173 Internal        OverallRR:100%       13          2hoursago
```

From Backend
```
ajesh@ajesh:~/litmus-test$ kubectl get pods -n litmus 
NAME                                                  READY   STATUS      RESTARTS   AGE
argo-server-5d995b5fcd-b5jmd                          1/1     Running     0          6h1m
argowf-chaos-kube-proxy-chaos-1604378173-1099108954   0/2     Completed   0          5h52m
argowf-chaos-kube-proxy-chaos-1604378173-229476964    0/2     Completed   0          5h52m
argowf-chaos-kube-proxy-chaos-1604378173-2425261330   0/2     Completed   0          5h50m
argowf-chaos-kube-proxy-chaos-1604378173-3450321278   0/2     Completed   0          5h50m
argowf-chaos-kube-proxy-chaos-1604378173-3896914393   0/2     Completed   0          5h48m
argowf-chaos-kube-proxy-chaos-1604378173-574301072    0/2     Completed   0          5h52m
argowf-chaos-kube-proxy-chaos-1604378173-594986771    0/2     Completed   0          5h46m
chaos-operator-ce-684c5544d9-wppn7                    1/1     Running     0          6h1m
litmusportal-frontend-686c8b878c-h2glt                1/1     Running     2          6h5m
litmusportal-server-54b47d5b4d-ls558                  2/2     Running     1          6h5m
mongo-5b58444bc5-qgl4s                                1/1     Running     0          6h5m
subscriber-74fb6bff7-59kg6                            1/1     Running     0          5h54m
workflow-controller-65759bf57f-sz4c2                  1/1     Running     0          6h1m
```

### PR DESCRIPTION
**Pull request by gdsoumya**
Signed-off-by: Soumya Ghosh Dastidar <gdsoumya@gmail.com>

## Proposed changes

fixes #2310 

## Types of changes

What types of changes does your code introduce to Litmus? Put an `x` in the boxes that apply
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices applies)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
- [x] I have read the [CONTRIBUTING](https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the commit for DCO to be passed.
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go [hunk 1]
```diff
@@ -44,32 +44,34 @@ func UpdateCluster(query bson.D, update bson.D) error {
...
 	opts := options.Update().SetUpsert(true)
 	ctx, _ := context.WithTimeout(backgroundContext, 10*time.Second)
 
 	count, err := workflowCollection.CountDocuments(ctx, bson.M{"workflow_id": workflow_id, "workflow_runs.workflow_run_id": wfRun.WorkflowRunID})
 	if err != nil {
-		return err
+		return 0, err
 	}
 
+	updateCount := 1
```
Comment (rajdas98) [inline litmus-portal/graphql-server/pkg/database/mongodb/operations.go:56]: In which case, update count will be 0?

File: litmus-portal/graphql-server/pkg/graphql/mutations/mutation.go [hunk 2]
```diff
@@ -123,16 +123,21 @@ func WorkFlowRunHandler(input model.WorkflowRunInput, r store.StateData) (string
...
 		ExecutionData: input.ExecutionData,
+		Completed:     input.Completed,
 	})
 	if err != nil {
 		log.Print("ERROR", err)
 		return "", err
 	}
 
+	if count == 0 {
+		return "Workflow Run Accepted", nil
```
Comment (rajdas98) [inline litmus-portal/graphql-server/pkg/graphql/mutations/mutation.go:138]: Can you change the return message? The message is the same if count != 0

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go [hunk 3]
```diff
@@ -44,32 +44,34 @@ func UpdateCluster(query bson.D, update bson.D) error {
...
 	if count == 0 {
 		filter := bson.M{"workflow_id": workflow_id}
 		update := bson.M{"$push": bson.M{"workflow_runs": wfRun}}
 		_, err = workflowCollection.UpdateOne(ctx, filter, update, opts)
 		if err != nil {
```
Comment (rajdas98) [inline litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]:
```suggestion
	updateCount := 0
	if count == 0 {
		filter := bson.M{"workflow_id": workflow_id}
		update := bson.M{"$push": bson.M{"workflow_runs": wfRun}}
		updateResp, err = workflowCollection.UpdateOne(ctx, filter, update, opts)
		if err != nil {
		    return 0, err
		}
		updateCount = int(updateResp.MatchedCount)
```

**Review by rajdas98 — CHANGES_REQUESTED** [review]

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 3 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: This is not needed, the updateCount is required for only already existing workflow_runs

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 1 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:56]: If the `completed` param is true and not false

File: litmus-portal/graphql-server/pkg/graphql/mutations/mutation.go (hunk 2 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/graphql/mutations/mutation.go:138]: Not needed, the subscriber as it is doesn't need to know about the duplicate events because it cannot do anything about these events. The events are generated by the workflow controller

Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/graphql/mutations/mutation.go]: Updated

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 3 above)
Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: In case of `count == 0`, it is also updated. It is required if updates fail during that time

Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: If the update fails we will return the error. If there is no error then that means the update succeeded the updateCount is irrelevant in that case. For count==0 case this patch is irrelevant altogether.

Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: If updateCount is specific for existing workflow_runs make it boolean. It should operate based on true/false.

Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]:
```
		if int(updateResp.MatchedCount) > 0 {
                         isExisting = true
                 }

```

Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: No, the logic is not about whether the workflow_run exists but whether a record is updated. If there is an update the updateCount will always be 1. If there is no update it will be 0.  That is why in the error paths I have mentioned updateCount as 0 because no record is being updated, besides returning the error. The default value is 1 because for count==0 the updateCount will always be 1 (or 0 in error) as only 1 workflow_run can been updated/inserted. I think keeping it integer suits the usecase and logic.

Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: > If the update fails we will return the error. If there is no error then that means the update succeeded the updateCount is irrelevant in that case. For count==0 case this patch is irrelevant altogether.

In some cases, MongoDB will not return err, if update query params didn't match(It will return matchcount == 0). In that case update count remains 1

Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: In count == 0, the updateCount will not always return 0 in err, as mention the case here- https://github.com/litmuschaos/litmus/pull/2312/files#r516842184

Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: That case wasn't handled previously? If this is a possibility it should've been handled. Though I doubt this is possible because the code always makes a db entry for the workflow first and only after that succeeds it sends the workflow to the agent. The wf_id if provided by the agent should always match.

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go [hunk 4]
```diff
@@ -44,32 +44,34 @@ func UpdateCluster(query bson.D, update bson.D) error {
...
-		update := bson.M{"$set": bson.M{"workflow_runs.$.last_updated": wfRun.LastUpdated, "workflow_runs.$.execution_data": wfRun.ExecutionData}}
-		_, err = workflowCollection.UpdateOne(ctx, filter, update, opts)
+		filter := bson.M{"workflow_id": workflow_id, "workflow_runs.workflow_run_id": wfRun.WorkflowRunID, "workflow_runs.completed": false}
+		update := bson.M{"$set": bson.M{"workflow_runs.$.last_updated": wfRun.LastUpdated, "workflow_runs.$.execution_data": wfRun.ExecutionData, "workflow_runs.$.completed": wfRun.Completed}}
+		updateResp, err := workflowCollection.UpdateOne(ctx, filter, update, opts)
 		if err != nil {
-			return err
+			return 0, err
 		}
+		updateCount = int(updateResp.MatchedCount)
```
Comment (rajdas98) [inline litmus-portal/graphql-server/pkg/database/mongodb/operations.go:71]: Also use `updateResp.Modified` instead  `updateResp.MatchedCount`

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 3 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: If that's a possibility we should be throwing an error when the wf_id doesn't match.

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 4 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:71]: I followed the documentation for UpdateOne() 
```
If the filter does not match any documents, the operation will succeed and an UpdateResult with a MatchedCount of 0 will be returned.
```

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 3 above)
Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: It wasn't handled previously, It is not about wf_id, if workflos_runs  is not available it will return `Modified: 0`

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 4 above)
Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:71]: I tried in my local system, the query is matched but I am updating a field which is not present.
it returns `MatchedCount: 1, Modified:0 `
```
> db.col.find()
{ "_id" : ObjectId("5fa19715e044d519bc185e97"), "id" : 2 }
> db.col.update({"id": 2}, {"idd": 2})
WriteResult({ "nMatched" : 1, "nUpserted" : 0, "nModified" : 1 })

```

Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:71]: We have to add both conditions.

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 3 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: Did not get you, the query is about workflow_id right, so the update is dependant on the workflow_id itself. The workflow_runs array can be empty but that shouldn't cause any problems as we are pushing a new workflow inside it. The update count should be 1 as long as the wf_id matches

Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: The thing is don't initialize updateCount == 1 at first, we can't predict the mongodb response.
Set to 1 once you got the response from the mongodb

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 4 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:71]: this is an impossible situation in our case because our apis and db schemas are strict. Neither can the agent send any data that isn't accounted for in the gql schema nor can it be marshalled to the db struct as it is also strongly typed.

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 3 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: Okay I'll update the code to throw an error if the matchCount is 0 for the wf_id. That should solve both the problems

Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:61]: Updated, it should now throw error if not workflow_id match is found.

File: litmus-portal/graphql-server/pkg/database/mongodb/operations.go (hunk 4 above)
Comment (rajdas98) [reply litmus-portal/graphql-server/pkg/database/mongodb/operations.go:75]: We can leave this part. we have to thing more edge case for the test cases.

**Review by rajdas98 — DISMISSED** [review]

**Comment by rajdas98** [comment]
If you test this PR, we can merge this

**Comment by gdsoumya** [comment]
> If you test this PR, we can merge this

I tested it, though couldn't test the actual bug because I can't seem to reproduce it in my system, will need help from @ajeshbaby  to verify that.

**Comment by rajdas98** [comment]
Can you build a custom image of this PR and give it to @ajeshbaby to test it ?

**Comment by gdsoumya** [comment]
> Can you build a custom image of this PR and give it to @ajeshbaby to test it ?

I manually re-created the situations and tested the code, it is working as expected. Here is the subscriber logs:

- Logs for workflows with no matching wf_id :

```
time="2020-11-04T02:56:18Z" level=info msg="WORKFLOW EVENT d2e410b2-48f4-46f9-9159-b845212d13c3 UPDATE"
time="2020-11-04T02:56:18Z" level=info msg="RESPONSE {\"errors\":[{\"message\":\"workflow not found\",\"path\":[\"chaosWorkflowRun\"]}],\"data\":null}"
time="2020-11-04T02:56:18Z" level=info msg="WORKFLOW EVENT d2e410b2-48f4-46f9-9159-b845212d13c3 UPDATE"
time="2020-11-04T02:56:18Z" level=info msg="RESPONSE {\"errors\":[{\"message\":\"workflow not found\",\"path\":[\"chaosWorkflowRun\"]}],\"data\":null}"
```

- Logs for duplicate updates [actual bug] :

```
time="2020-11-04T03:04:24Z" level=info msg="RESPONSE {\"data\":{\"chaosWorkflowRun\":\"Workflow Run Discarded[Duplicate Event]\"}}"
```
